src/device_manager.rs

```diff
@@ -0,0 +1,431 @@
+    /// Address allocation failed.
+    AddressAllocate(AllocatorError),
+}
+
+/// Simplify the `Result` type.
+pub type Result<T> = result::Result<T, Error>;
+
+/// System device manager serving for all devices management and VM exit handling.
+pub struct DeviceManager {
+    /// System allocator reference.
+    resource: Arc<Mutex<SystemAllocator>>,
+    /// Devices information mapped by instance id.
+    devices: HashMap<u32, DeviceDescriptor>,
+    /// Range mapping for VM exit mmio operations.
+    mmio_bus: BTreeMap<Range, Arc<Mutex<dyn Device>>>,
+    /// Range mapping for VM exit pio operations.
+    pio_bus: BTreeMap<Range, Arc<Mutex<dyn Device>>>,
+}
+
+impl DeviceManager {
+    /// Create a new `DeviceManager` with a `SystemAllocator` reference which would be
+    /// used to allocate resource for devices.
+    pub fn new(resource: Arc<Mutex<SystemAllocator>>) -> Self {
+        DeviceManager {
+            resource,
+            devices: HashMap::new(),
+            mmio_bus: BTreeMap::new(),
+            pio_bus: BTreeMap::new(),
+        }
+    }
+
+    fn insert(&mut self, dev: DeviceDescriptor) -> Result<()> {
+        // Insert if the key is non-present, else report error.
+        if self.devices.get(&(dev.instance_id)).is_some() {
+            return Err(Error::Exist);
+        }
+        self.devices.insert(dev.instance_id, dev);
+        Ok(())
+    }
+
+    fn remove(&mut self, instance_id: u32) -> Option<DeviceDescriptor> {
+        self.devices.remove(&instance_id)
+    }
+
+    fn device_descriptor(
+        &self,
+        id: u32,
+        name: String,
+        dev: Arc<Mutex<dyn Device>>,
+        parent_bus: Option<Arc<Mutex<dyn Device>>>,
+        resources: Vec<IoResource>,
+        irq: Option<IrqResource>,
+    ) -> DeviceDescriptor {
+        DeviceDescriptor::new(id, name, dev.clone(), parent_bus, resources, irq)
+    }
+
+    // Allocate IO and instance id resources.
+    // Return a Result wrapper of instance id.
+    fn allocate_resources(&mut self, resources: &mut Vec<IoResource>) -> Result<(u32)> {
+        let id = self
+            .resource
+            .lock()
+            .expect("failed to acquire lock.")
+            .allocate_instance_id()
+            .map_err(Error::InstanceIdAllocate)?;
+
+        for res in resources.iter_mut() {
+            match res.res_type {
+                IoType::Pio => {
+                    if res.addr.is_none() {
+                        return Err(Error::NonePIORequest);
+                    }
+                    res.addr = Some(
+                        self.resource
+                            .lock()
+                            .expect("failed to acquire lock")
+                            .allocate_io_addresses(res.addr.unwrap(), res.size)
+                            .map_err(Error::AddressAllocate)?,
+                    );
+                }
+                IoType::PhysicalMmio | IoType::Mmio => {
+                    res.addr = Some(
```

Comment:
We need a flag to mark whether the resource is dynamically allocated from the resource pool or statically reserved. So we could do the right thing when frees the resource.

Reply:
Do you mean a specified address request and non-specified one? If so, both of them are being allocated and marked there. And both of them can be freed when unregister_device.

Reply:
> We need a flag to mark whether the resource is dynamically allocated from the resource pool or statically reserved.

What do you mean by statically reserved? Physical RAM for example?

Reply:
> > Seems more thoughts are needed about the DeviceManager abstraction. Especially the way to support both simple VMM like firecracker and full functionally vmm like qemu
> 
> The current design is driven by the assumption that the VMM will be the central resource allocation point, which is true for firmware-less and simple VMMs like e.g. Firecracker or to some extend crosvm. It's also a little static at the moment as it does not yet allow for device resources reprogramming (e.g. PCI).
> However, I think this is a good starting point and it conveys the design of what both Firecracker and crosvm do. From there, we can extend the device manager API to support more complex use cases.
> 
> > they have different requirements.
> 
> Maybe we could try to enumerate those requirements and see if the current PR could eventually also handle them. Besides device driven resources re-allocation (aka reprogramming), what else do you think we should take into account?

Sure, we focus on simple static alike VMM too:)

Reply:
> > We need a flag to mark whether the resource is dynamically allocated from the resource pool or statically reserved.
> 
> What do you mean by statically reserved? Physical RAM for example?

Actually I hope to make free_resources() undo all changes made by allocate_resources().
For an IoResource entry 'res', allocate_resoures() will set res.addr to Some(addr) on success when called with res.addr as None. So I hope res.addr will be reset to None after calling free_resources().

Reply:
> > > We need a flag to mark whether the resource is dynamically allocated from the resource pool or statically reserved.
> > 
> > 
> > What do you mean by statically reserved? Physical RAM for example?
> 
> Actually I hope to make free_resources() undo all changes made by allocate_resources().
> For an IoResource entry 'res', allocate_resoures() will set res.addr to Some(addr) on success when called with res.addr as None. So I hope res.addr will be reset to None after calling free_resources().

Thanks for the detailed explanation.  Referring to [above comment](https://github.com/rust-vmm/vm-device/pull/3#discussion_r300038235), I' thinking if it is really need to set original resource status back to device instance when unhotplug device. Because device instances are managed by DeviceManager and all operations are not able to be executed after unregister_device(). And next time, when re-hotplug this device, it will re-tell what kinds of resouces it needs.